spotbugs/src/main/java/edu/umd/cs/findbugs/detect/CompoundOperationsOnSharedVariables.java

```diff
@@ -0,0 +1,215 @@
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ */
+
+package edu.umd.cs.findbugs.detect;
+
+import edu.umd.cs.findbugs.BugAccumulator;
+import edu.umd.cs.findbugs.BugInstance;
+import edu.umd.cs.findbugs.BugReporter;
+import edu.umd.cs.findbugs.ba.ClassContext;
+import edu.umd.cs.findbugs.ba.XField;
+import edu.umd.cs.findbugs.ba.XMethod;
+import edu.umd.cs.findbugs.bcel.OpcodeStackDetector;
+import edu.umd.cs.findbugs.util.MultiThreadedCodeIdentifierUtils;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import org.apache.bcel.Const;
+import org.apache.bcel.classfile.JavaClass;
+import org.apache.bcel.classfile.Method;
+
+public class CompoundOperationsOnSharedVariables extends OpcodeStackDetector {
+
+    private static final List<Short> readOpCodes = Arrays.asList(Const.GETFIELD, Const.GETSTATIC,
+            Const.ALOAD, Const.ALOAD_0, Const.ALOAD_1, Const.ALOAD_2, Const.ALOAD_3,
+            Const.DLOAD, Const.DLOAD_0, Const.DLOAD_1, Const.DLOAD_2, Const.DLOAD_3,
+            Const.LLOAD, Const.LLOAD_0, Const.LLOAD_1, Const.LLOAD_2, Const.LLOAD_3,
+            Const.FLOAD, Const.FLOAD_0, Const.FLOAD_1, Const.FLOAD_2, Const.FLOAD_3,
+            Const.ILOAD, Const.ILOAD_0, Const.ILOAD_1, Const.ILOAD_2, Const.ILOAD_3,
+            Const.DALOAD, Const.LALOAD, Const.FALOAD, Const.IALOAD);
+
+    private static final List<Short> pushOpCodes = Arrays.asList(Const.DCONST_0, Const.DCONST_1,
+            Const.LCONST_0, Const.LCONST_1,
+            Const.FCONST_0, Const.FCONST_1, Const.FCONST_2,
+            Const.ICONST_0, Const.ICONST_1, Const.ICONST_2, Const.ICONST_3, Const.ICONST_4, Const.ICONST_5,
+            Const.LDC, Const.LDC_W, Const.LDC2_W);
+
+    private static final List<Short> possibleCompoundOperationOpCodes = Arrays.asList(
+            // +=,++,       -=,--       *=,         /=,         %=
+            Const.DADD, Const.DSUB, Const.DMUL, Const.DDIV, Const.DREM,
+            Const.FADD, Const.FSUB, Const.FMUL, Const.FDIV, Const.FREM,
+            Const.LADD, Const.LSUB, Const.LMUL, Const.LDIV, Const.LREM,
+            Const.IADD, Const.ISUB, Const.IMUL, Const.IDIV, Const.IREM,
+            // <<=,         >>=,        >>>=
+            Const.ISHL, Const.ISHR, Const.IUSHR,
+            Const.LSHL, Const.LSHR, Const.LUSHR,
+            // &=
+            Const.IAND, Const.LAND,
+            // |=, ^=
+            Const.IOR, Const.IXOR, Const.LOR, Const.LXOR);
+
+    private final BugAccumulator bugAccumulator;
+    private final Map<XMethod, List<XField>> compoundlyWrittenNotSecuredFieldsByMethods;
+    private final Map<XMethod, List<XField>> readNotSecuredFieldsByMethods;
+    private boolean isInsideSynchronizedOrLockingMethod;
+    private Optional<XField> maybeCompoundlyOperatedField;
```

Comment:
In my understanding, it is not suggested to use `Optional` as type of fields.

* https://dzone.com/articles/optional-in-java
* https://www.baeldung.com/java-optional
* (I cannot find it but I remember that Oracle's official tutorial had similar suggestions)

Reply:
I understand, changed it to use a `null` default value.